### PR TITLE
[api-extractor] add basic support for `import * as module from './local-module'`

### PR DESCRIPTION
**Pull request by mckn**
I have basically taken the PR created by @adventure-yunfei here PR and cleaned up the code a bit. Tested it on the Grafana project which is a really big project and then opened this PR.

Thanks and great work by @adventure-yunfei!

Please let me know what you want to be adressed to get it merged. I think there are several people that wants to help out to get this PR merged.

Fixes: #1029

### CODE REVIEW

**Comment by stevengum** [comment]
@octogonz sorry for the out-of-the-blue PR review, just excited to use `api-extractor` on my team's projects 😄 

@mckn, do you have tests that can be added to the PR?

**Comment by mckn** [comment]
> @octogonz sorry for the out-of-the-blue PR review, just excited to use `api-extractor` on my team's projects 😄
> 
> @mckn, do you have tests that can be added to the PR?

@stevengum awesome feedback! Thanks for doing the review. Have made the updates and pushed the changes.

I have not added any tests but please give me some directions on how I can write some tests and I am happy to add that as well.

**Comment by stevengum** [comment]
@mckn sorry for the delay, I'm not a regular contributor to this project so I don't know how to write/run tests. 

I've pulled the grafana branch and was going to run the tests locally, it looks like they've already passed as part of the setup of the repo. 

I don't have insight into their CI pipelines either, could the `rushstack CI Build` be stuck in pending due to the merge conflict?

**Comment by mckn** [comment]
@stevengum I have added one some simple tests to verify the `import * as ..` scenario. Not sure if it is the merge conflict that is causing errors. But I will sync with upstream asap so we can get this mergeable.

**Comment by mckn** [comment]
@octogonz would be awesome if you could have a look at this PR 🙏

**Comment by rbuckton** [comment]
I tried this on one of my projects and the links in the TOC for namespaces are broken.

**Comment by rbuckton** [comment]
Disregard that comment. After a clean build of my documentation it seems to be working fine.

**Comment by rbuckton** [comment]
I will add that this does fix the underlying issue for me. I've been using a local patch to work around this for my documentation builds and that patch is no longer necessary.

**Comment by mckn** [comment]
The question is: what can I do to get this one merged?

File: apps/api-extractor/src/analyzer/AstImportAsModule.ts
Comment (rbuckton) [inline apps/api-extractor/src/analyzer/AstImportAsModule.ts:12]: Was there a reason to introduce a new class here rather than integrate this capability into `AstSymbol`?

Comment (rbuckton) [reply apps/api-extractor/src/analyzer/AstImportAsModule.ts:12]: Or the existing `AstImport`, which has `AstImportKind.StarImport`?

File: apps/api-extractor/src/analyzer/AstSymbolTable.ts
Comment (rbuckton) [inline apps/api-extractor/src/analyzer/AstSymbolTable.ts:149]: You've opened `AstSymbolTable.analyze` to accept `AstEntity` which includes `AstImport`, but there are no cases that handle it in this function.

**Comment by octogonz** [comment]
> The question is: what can I do to get this one merged?

I promise I'll review it tomorrow. 😇

**Comment by octogonz** [comment]
@mckn could you unlock your branch so I can push updates to it?  I'd like to fix a few things.

**Comment by octogonz** [comment]
@rbuckton Wrote:
> Was there a reason to introduce a new class here rather than integrate this capability into `AstSymbol`?

This is an important question.  In API Extractor 7, we eliminated all the special types (`AstClass`, `AstFunction`, etc.) and replaced them with a generalized `AstSymbol` that corresponds to a compiler symbol, and has associated `AstDeclaration` objects that correspond essentially to spans of text in the .d.ts files.  This simplified API Extractor by delegating substantial work to the compiler and and even to the .d.ts content.  Besides simplifying the code, this approach makes API Extractor very robust against TypeScript compiler syntax changes.  (For example, although API Extractor 
 relies on TypeScript 3.7 which emits [incompatible getters/setters](https://github.com/microsoft/TypeScript/issues/33939), API Extractor will nonetheless generate correct rollups for older .d.ts files because it doesn't rely on the compiler's emitter.)

However, `AstSymbol` isn't good for representing an entity that doesn't have a meaningful compiler symbol+declaration.  Today there's only one such case:

```ts
export type AstEntity = AstSymbol | AstImport;
```

Although imports do have compiler symbols, the `AstImport` is a mostly synthetic concept because:
- It gets emitted in a purely custom way that does not rely on the input .d.ts syntax tree
- The information that it stores is very specialized, and largely unrelated to the compiler's analysis
- Today it doesn't get emitted in .api.json file (but that was a mistake)

With `AstImport` alone we managed to handle four different cases of `AstImportKind` reasonably well.

But `import * as ___` stretches the requirements substantially, because it creates a synthetic `namespace` that contains other `AstEntity` members. I can see why @adventure-yunfei chose to add a third class `AstImportAsModule`.  It's blazing the trail for other synthetic constructs, which is good!

But we need to stop and think about the right class design, since if we open up `AstEntity` then every step of the engine needs to understand the new possibilities.  The includes some loose ends that aren't addressed yet in this PR:

- It doesn't write the API report
- While it emits a `namespace` in the .api.json file, the signature is missing (`"docComment"` field)

So I'd like to mull over the architecture a bit before proceeding.  The PR itself is fortunately is not very big, so whichever way we go should be relatively quick to code.

**Comment by octogonz** [comment]
Some approaches I'm comparing are:

1. Convert `AstEntity` into an abstract base class that has children like `AstSymbol`, `AstNamedImport`, `AstSimpleImport`, and now `AstImportAsModule`

2. Have fundamentally two kinds of `AstEntity`:  `AstSymbol` (which regurgitates hunks of code from the .d.ts file) versus `AstSyntheticEntity` (which gets emitted in a custom way).  Then have the special cases be subclasses of `AstSyntheticEntity`

3. Somehow tie the `AstFoo` taxonomy more closely to the `ApiFoo` taxonomy from `api-extractor-module`.  It's slightly weird that these two worlds are so separate.

At first glance #2 seems most promising...

**Comment by rbuckton** [comment]
In the TS compiler we would (and do) just use a `Symbol` differentiated by a `SymbolFlags` value.

**Comment by adventure-yunfei** [comment]
@octogonz agree that keep using `AstSymbol` for `import * as ___` is perfect.

At first glance I thought it should be that way, since ts compiler gives us the symbol and the declaration for that: ts.SourceFile. But with my invesitigation, `AstSymbol` has some constraints under current implementation, while ts.SourceFile doesn't fit to those.

I've tried to make ts.SourceFile as a supported delcaration kind but failed. There're too many related designs. Perhaps there's a way to update those contraints and treat ts.SourceFile and other declarations in the same way.

**Comment by mckn** [comment]
> @mckn could you unlock your branch so I can push updates to it? I'd like to fix a few things myself to get this merged faster.

That would be great! But not really sure how to do that 🙈

**Comment by mckn** [comment]
> @mckn could you unlock your branch so I can push updates to it? I'd like to fix a few things myself to get this merged faster.

An alternative is that you open a PR to this branch and I will merge it directly.

**Comment by octogonz** [comment]
@mckn when you first create the PR, you'll see a checkbox like this:


> allow-edits-1

Afterwards, that checkbox is moved to  the corner of the page like this:


> allow-edits-2

**Comment by mckn** [comment]
> @mckn when you first create the PR, you'll see a checkbox like this:
> 
> > allow-edits-1
> 
> Afterwards, that checkbox is moved to the corner of the page like this:
> 
> > allow-edits-2

hmm, I don't have that option available. Trying to find out why.

Skärmavbild 2020-04-09 kl  08 39 36

**Comment by mckn** [comment]
@octogonz you should now be able to do some changes in the repo

**Comment by mckn** [comment]
@octogonz what time zone do you work in? Maybe we can try find a overlapping time and try to make the changes together to get this one merged.

**Comment by octogonz** [comment]
I'm in PST and usually reachable on [Gitter](https://gitter.im/octogonz) or Twitter for DMs. (Right now everything is in slow motion due to the lockdown, though, so I apologize if responses are lagged a bit.)

**Comment by octogonz** [comment]
> octogonz you should now be able to do some changes in the repo

🤔 It's still not working. 

I will just create a PR into your PR and work from that.  

PR https://github.com/microsoft/rushstack/pull/1833

**Comment by octogonz** [comment]
(I figured out the branch permissions; I had to click "accept" on an invitation email from GitHub.)

**Comment by mckn** [comment]
@octogonz let me know if I can help out fixing things :)

**Comment by mckn** [comment]
What do I need to update to get this PR mergeable? Feels like it have been stale for a long time now.

**Comment by octogonz** [comment]
> What do I need to update to get this PR mergeable? Feels like it have been stale for a long time now. Starting to think it might be faster to publish my own version of the api-extractor.

I've been swamped lately. This PR is at the top of my list, but there's some missing pieces that need to be completed before we can merge it. 

If it would unblock you, we can publish a "prerelease" version of the current PR branch.

If someone wants to help push it along, the missing pieces are:
- Add the code to emit the namespace in the the API report (instead of reporting an error `'"import * as ___ from ___;" is not supported for local files`)
- In the .api.json file, add the missing `"docComment"` field
- Refactor `AstEntity` as described above; I probably have to do this one. It's not a lot of work, but I need to find an hour of completely uninterrupted time to think through the design consequences. :-)

Really sorry about the delays. Life is moving in slow motion with this lockdown...

**Comment by octogonz** [comment]
> If it would unblock you, we can publish a "prerelease" version of the current PR branch.

I published a beta release of API Extractor as version `7.8.2-pr1796.0`, with the current contents of this branch.

I spent some more time thinking about how to generalize the design of `AstEntity`, and I also looked at the `SymbolFlags` pattern suggested by @rbuckton.  I'll follow up when I have a working prototype.

**Comment by mckn** [comment]
> > If it would unblock you, we can publish a "prerelease" version of the current PR branch.
> 
> I published a beta release of API Extractor as version `7.8.2-pr1796.0`, with the current contents of this branch.
> 
> I spent some more time thinking about how to generalize the design of `AstEntity`, and I also looked at the `SymbolFlags` pattern suggested by @rbuckton. I'll follow up when I have a working prototype.

Awesome! Great work @octogonz! Please let me know if I can help out in any way :)

**Comment by octogonz** [comment]
FYI we recently reformatted the master branch using Prettier, which creates merge conflicts for existing PRs. We'll try to fix up the existing PRs for people, but if you want to do it yourself, here's the steps:
 
1. Run "npm install -g prettier@2.0.5" to install Prettier.
2. Merge the "release/master-before-prettier" branch into your PR branch. This is the state of the world before Prettier was applied.
3. Merge the "release/master-after-prettier" branch into your PR branch. This will create a merge conflict with the Prettier changes, and nothing else.
4. Resolve all the merge conflicts by discarding "origin", and overwriting with your branch's changes.
5. BEFORE committing the merge, run "prettier . --write" from the repo root to reapply the formatting (that you discarded in step 4).
6. Make sure to restage everything.
7. Commit the merge
8. Now you should be able to merge from the real "master" without any (Prettier-related) merge conflicts

**Comment by ctaggart** [comment]
I think I just [hit this problem](https://github.com/Azure/azure-sdk-for-js/pull/9502#issuecomment-646232849), trying to extract the API for an azure service sdk. It would be great to have this fixed and merged in soon.

**Comment by clar-cmp** [comment]
Would also love to see this merged. Any plans to get to this soon?

**Comment by octogonz** [comment]
> Would also love to see this merged. Any plans to get to this soon?

Yes, this and #1916 should both get some attention soon. They are related.

**Comment by mckn** [comment]
@octogonz let me know if I can help out in some way.

**Comment by octogonz** [comment]
> let me know if I can help out in some way.

Will do. I made a bunch of progress with this, however the "todo" parts (API reporting and .api.json generation) raise some more questions about the best way to model these synthetic constructs. We don't necessarily have to implement that in this PR, but I want to at least come up with a design to make sure the refactor is on the right track.

**Comment by mckn** [comment]
We have upgraded our project to typescript `4.0.2` and we get the following errors when running the `7.8.2-pr1796.0` version:

`Error: Symbol not found for identifier: FlatArray`

Do you think it will be resolved by merging the latest version of API extractor into this PR and publish a new version of this PR?

**Comment by mckn** [comment]
> @octogonz
> 
> We have upgraded our project to typescript `4.0.2` and we get the following errors when running the `7.8.2-pr1796.0` version:
> 
> `Error: Symbol not found for identifier: FlatArray`
> 
> Do you think it will be resolved by merging the latest version of API extractor into this PR and publish a new version of this PR?

It did fix the issue I had but looks like we need to get this one in place as well: https://github.com/microsoft/rushstack/pull/1916

**Comment by Jack-Works** [comment]
Hi, hope to see this shipped. All of our main API is exported like:

```ts
export * as NS1 from './path1'
export * as NS2 from './path2'
```

**Comment by octogonz** [comment]
Some Microsoft people are asking for this as well now. 

I'm going to move this PR and https://github.com/microsoft/rushstack/pull/1916 to the top of my list.  It's embarrassing how long these important PRs have been open. 😅

**Comment by santam85** [comment]
Happy to see there's a PR already in this advanced stage for this. This is a must to support api-extractor for Angular projects as well, as there's limited control on the typings output from the CLI

**Comment by lifeiscontent** [comment]
@octogonz any updates here?

**Comment by octogonz** [comment]
> @octogonz any updates here?

To be clear, what's blocking this PR wasn't the code review, but rather some nontrivial work required to complete the implementation.  (Some of which is identified by `TODO` in the author's comments.)  I'm looking at it again this weekend to see if I can push it along.

**Comment by octogonz** [comment]
Today I got the `.api.md` file straightened out. The last step will be the `.api.json` file.

**Comment by octogonz** [comment]
🚀 This is finally ready to go! 

I tested it by building a large production monorepo with lots of projects that use API Extractor.

**Review by octogonz — APPROVED** [review]

**Comment by octogonz** [comment]
Released with API Extractor 7.17.0

**Pull request merged**
